worker-message-prioritization-rerouting/src/main/java/com/github/workerframework/workermessageprioritization/rerouting/MessageRouterSingleton.java

```diff
@@ -126,19 +126,36 @@ public static void route(final Document document) {
             try {
                 messageRouter.route(document);
             } catch (final Throwable throwable) {
-                LOGGER.error("Exception thrown trying to route document", throwable);
-
-                // If an error has been thrown by the messageRouter.route call, it is possible that the connection and/or channel has
-                // been closed, so we should attempt to init this class again the next time init() is called to create a new connection
-                // and new channel.
-                closeQuietly();
-                initShouldBeReattempted = true;
+                handleRouteFailure(throwable);
+                throw throwable;
+            }
+        }
+    }
 
+    public static String route(final String originalQueueName, final String tenantId) {
+        if(messageRouter != null) {
+            try {
+                return messageRouter.route(originalQueueName, tenantId);
+            } catch (final Throwable throwable) {
+                handleRouteFailure(throwable);
                 throw throwable;
             }
+        } else {
+            return originalQueueName;
         }
     }
 
+    private static void handleRouteFailure(final Throwable throwable) {
+        LOGGER.error("Exception thrown trying to route document", throwable);
+
+        // If an error has been thrown by the messageRouter.route call, it is possible that the connection and/or channel has
```

Comment:
This feels clunky as we don't know if the reason for the failure is something that will be fixed by a re-init.
Would it be possible for the Router to handle reconnecting so that throwables reaching this level can always be handled as permanent failures?

Reply:
I've moved this recovery code to the StagingQueueCreator class now.